### PR TITLE
【Hackathon 6th No.11】为 Paddle 新增 bernoulli_  API - part

### PR DESCRIPTION
**Pull request by NKNaN**
### PR Category

User Experience

### PR Types

New features

### Description

新增 bernoulli_ API

### CODE REVIEW

File: python/paddle/tensor/random.py [hunk 1]
```diff
@@ -104,6 +104,48 @@ def bernoulli(x, name=None):
...
+             [0., 1., 1., 0.],
+             [0., 1., 1., 1.]])
+
+    """
+    x.uniform_(0.0, 1.0)
+    ones_mask = x > p
+    zeros_mask = x < p
+    x.masked_fill_(ones_mask, 1.0)
+    out = x.masked_fill_(zeros_mask, 0.0)
+    return out
```
Comment (zhwesky2010) [inline python/paddle/tensor/random.py]: x.masked_fill_(ones_mask, 1.0)
x.masked_fill_(zeros_mask, 0.0)
return x

这样吧

File: test/deprecated/legacy_test/test_inplace.py [hunk 2]
```diff
@@ -1921,5 +1921,48 @@ def test_inplace_api(self):
...
+    def init_data(self):
+        self.shape = (100, 1000)
+        self.input_var_numpy = np.random.random(self.shape)
+        self.dtype = "float32"
+        self.p = 0.5
+
+    def set_np_compare_func(self):
+        self.np_compare = np.array_equal
+
+    def inplace_api_processing(self, var):
```
Comment (zhwesky2010) [inline test/deprecated/legacy_test/test_inplace.py:1938]: 也测一下类方法调用吧，x.bernoulli_

Comment (NKNaN) [reply test/deprecated/legacy_test/test_inplace.py:1938]: 已增加

File: python/paddle/tensor/random.py (hunk 1 above)
Comment (NKNaN) [reply python/paddle/tensor/random.py]: 已修改

File: test/deprecated/legacy_test/test_inplace.py [hunk 3]
```diff
@@ -1921,5 +1921,65 @@ def test_inplace_api(self):
...
+        return paddle.bernoulli_(var, p=self.p)
+
+    def inplace_class_method_processing(self, var):
+        return var.bernoulli_(self.p)
+
+    def non_inplace_api_processing(self, var):
+        return paddle.bernoulli(paddle.zeros(self.shape) + self.p)
```
Comment (zhwesky2010) [inline test/deprecated/legacy_test/test_inplace.py]: 这个改成 paddle.full(self.shape, p) 更好

File: test/deprecated/legacy_test/test_inplace.py [hunk 4]
```diff
@@ -1921,5 +1921,65 @@ def test_inplace_api(self):
...
+
+    def test_inplace_api(self):
```
Comment (zhwesky2010) [inline test/deprecated/legacy_test/test_inplace.py:1947]: 测一下反向，随机数反向应该全0

File: test/deprecated/legacy_test/test_inplace.py [hunk 5]
```diff
@@ -1921,5 +1921,65 @@ def test_inplace_api(self):
...
+        np.testing.assert_allclose(
+            non_inplace_var.numpy().mean(),
+            inplace_var.numpy().mean(),
+            atol=0.01,
+        )
+        np.testing.assert_allclose(
+            non_inplace_var.numpy().var(), inplace_var.numpy().var(), atol=0.01
+        )
+
+    def test_inplace_class_method(self):
```
Comment (zhwesky2010) [inline test/deprecated/legacy_test/test_inplace.py:1973]: 测一下反向，随机数反向应该全0

File: test/deprecated/legacy_test/test_inplace.py (hunk 3 above)
Comment (NKNaN) [reply test/deprecated/legacy_test/test_inplace.py]: 已修改

File: test/deprecated/legacy_test/test_inplace.py (hunk 4 above)
Comment (NKNaN) [reply test/deprecated/legacy_test/test_inplace.py:1947]: 已添加反向测试

File: test/deprecated/legacy_test/test_inplace.py (hunk 5 above)
Comment (NKNaN) [reply test/deprecated/legacy_test/test_inplace.py:1973]: 已添加反向测试

**Comment by zhwesky2010** [comment]
@NKNaN 提交了吗

**Comment by NKNaN** [comment]
> @NKNaN 提交了吗

已提交

**Review by zhwesky2010 — DISMISSED** [review]
LGTM

File: python/paddle/tensor/random.py [hunk 6]
```diff
@@ -104,6 +104,48 @@ def bernoulli(x, name=None):
...
+@dygraph_only
+def bernoulli_(x, p=0.5, name=None):
+    """
+    This is the inplace version of api ``bernoulli``, which returns a Tensor filled
+    with random values sampled from a bernoulli distribution. The output Tensor will
+    be inplaced with input ``x``. Please refer to :ref:`api_tensor_bernoulli`.
+
+    Args:
+        x(Tensor): The input tensor to be filled with random values.
+        p (float|Tensor, optional): The success probability parameter of the output Tensor's bernoulli distribution.
```
Comment (jeff41404) [inline python/paddle/tensor/random.py:116]: Should we support scenarios where `p` can broadcast to `x`? if so, we should add description `p` must be broadcast to `x` here, and add unit tests to check error if cannot broadcast. If not, it needs to be clarified that p can only be float and Tensor which must have same shape with `x`, and add  unit tests to check error if `p.shape` != `x.shape`

Comment (NKNaN) [reply python/paddle/tensor/random.py:116]: We could support the scenarios. Let me update the code.

**Review by jeff41404 — APPROVED** [review]
LGTM

File: python/paddle/tensor/random.py [hunk 7]
```diff
@@ -104,6 +104,58 @@ def bernoulli(x, name=None):
         return out
 
 
+@dygraph_only
+def bernoulli_(x, p=0.5, name=None):
+    """
+    This is the inplace version of api ``bernoulli``, which returns a Tensor filled
+    with random values sampled from a bernoulli distribution. The output Tensor will
+    be inplaced with input ``x``. Please refer to :ref:`api_tensor_bernoulli`.
```
Comment (sunzhongkai588) [inline python/paddle/tensor/random.py:112]:
```suggestion
    be inplaced with input ``x``. Please refer to :ref:`api_paddle_bernoulli`.
```

File: python/paddle/tensor/random.py [hunk 8]
```diff
@@ -104,6 +104,58 @@ def bernoulli(x, name=None):
...
+    Returns:
+        A Tensor filled with random values sampled from the bernoulli distribution with success probability ``p`` .
+
+    Examples:
+        .. code-block:: python
+
+            >>> import paddle
+            >>> x = paddle.randn([3, 4])
+            >>> x.bernoulli_()
+            >>> # doctest: +SKIP('random check')
```
Comment (sunzhongkai588) [inline python/paddle/tensor/random.py:133]: 输出能固定吗？通过固定随机数

**Review by sunzhongkai588 — APPROVED** [review]
文档问题重新提一个 PR 改吧，这个先 approve

**Pull request merged**